gensim/models/fast_knn.py

```diff
@@ -0,0 +1,224 @@
+
+        return sorted(dict(wcd_distances).items(), key = itemgetter(1))
+
+    def get_wcd(self, test_doc, doc, doc_id):
+    #         print self.word_embedding.shape
+    #         print doc1.shape
+    #         print doc2.shape
+        """
+        This function calculates the Word Centroid Distance between docs
+        
+        Parameters
+        ----------
+
+        test_doc : Normalised BOW representaion of test doc .
+        doc : Normalised BOW representaion of stored doc .
+        doc_id : id of the stored doc
+        """
+        return (doc_id, distance.euclidean(np.dot(np.transpose(self.word_embedding), np.transpose(test_doc)), np.dot(np.transpose(self.word_embedding), doc)))
+    
+    def get_rwmd(self, test_doc, docs, doc_id):
+        """
+        This function calculates and returns the Relaxed Word Movers Distance (RWMD).
+
+        Parameters:
+        ----------
+
+        test_doc : Normalised BOW representaion of test doc .
+        doc : Normalised BOW representaion of stored doc .
+        doc_id : id of the stored doc
+        """
+        nonzeros_test_doc = np.nonzero(test_doc)[0]
+        nonzeros_train_doc = np.nonzero(docs[doc_id])[0]
+        union_bow = np.union1d(nonzeros_test_doc, nonzeros_train_doc)
+        w2v_distances = self.word_embedding_distance[np.ix_(union_bow, union_bow)]
+        np.fill_diagonal(w2v_distances, float('inf'))
+        dist1 = np.dot(np.transpose(test_doc)[union_bow][:,0], np.min(w2v_distances, axis = 1))
+        dist2 = np.dot(np.transpose(docs[doc_id][union_bow]), np.min(w2v_distances, axis = 1))
+        return (doc_id, max(dist1, dist2))
+
+    # def get_wmd():
+    #     wcd_distances = 
+    def get_wmd(self, test_doc, docs, doc_id):
+        """
+        This function calculates the wmd distances between 2 documents using emd function in pyemd
+
+        Parameters:
+        -----------
+
+        test_doc : Normalised BOW representaion of test doc .
+        doc : Normalised BOW representaion of stored doc .
+        doc_id : id of the stored doc
+        """
+        nonzeros_test_doc = np.nonzero(test_doc)[0]
+        nonzeros_train_doc = np.nonzero(docs[doc_id])[0]
+        union_bow = np.union1d(nonzeros_test_doc, nonzeros_train_doc)
+        # print union_bow.shape
+        # print test_doc.shape
+        # print docs[doc_id].shape
+        # print self.word_embedding_distance[np.ix_(union_bow, union_bow)].shape
+        return (doc_id,emd(float64(np.transpose(test_doc))[union_bow].ravel(), float64(docs)[doc_id][union_bow].ravel(), float64(self.word_embedding_distance[np.ix_(union_bow, union_bow)])))
+
+    def prune(self, wmd_dist_lst):
+        """
+        This function helps to place the new neighbor document at its right place.
+
+        Parameters:
+        -----------
+
+        wmd_dist_lst : list of wmd distances of n_neighbours from the test document
+        """
+        for doc_idx in range(self.n_neighbours-2,-1,-1):
```

Comment:
I will probably add comments in the next commit asap.
But, what I am essentially trying to do here is the following:
 wmd_dist_lst is a sorted array of WMD distances. So, when I detect that if some RWMD is less than the last element I get wmd distance for that document and substitute the last element  in the __getitems__ function if it is smaller than last  element and then try to move it to its rightful place by starting from neighbors-2 index till 0. 
If this is still not clear then I will refactor it. 
Please comment.